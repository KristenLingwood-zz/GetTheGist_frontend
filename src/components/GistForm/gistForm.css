.gist_form {
  margin-top: 5%;
  border-radius: 2%;
}

.description {
  margin-bottom: 2%;
}

.filename {
  margin-bottom: 1%;
}

button {
  float: right;
}
