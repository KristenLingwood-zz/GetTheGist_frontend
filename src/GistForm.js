import React, { PureComponent } from 'react';
import { FormControl, FormGroup, Button } from 'react-bootstrap';
import { callAPI } from './services/api';
import './gistForm.css';

class GistForm extends PureComponent {
  state = {
    description: '',
    filename: '',
    content: '',
    loading: false,
    fireRedirect: false
  };

  handleChange = e => {
    this.setState({ ...this.state, [e.target.name]: e.target.value });
  };

  handleNewSubmit = async e => {
    let payload = {
      description: this.state.description,
      filename: this.state.filename,
      content: this.state.content,
      extension: ''
    };
    try {
      await callAPI('post', 'http://localhost:3000/gists', payload);
      window.alert('Thanks for creating a new gist. We got the gist of it!');
      this.setState({
        description: '',
        filename: '',
        content: '',
        fireRedirect: true
      });
    } catch (err) {
      console.log(err);
    }
  };
  handleUpdateSubmit = async e => {
    console.log('update clicked');

    let payload = {
      description: this.state.description || this.props.foundGist.description,
      filename: this.state.filename || this.props.foundGist.filename,
      content: this.state.content || this.props.foundGist.content,
      extension: ''
    };
    try {
      let gistID = this.props.gistID;
      console.log(gistID);
      await callAPI('patch', `http://localhost:3000/gists/${gistID}`, payload);
      window.alert('Your gist has been updated!');
      this.setState({
        description: '',
        filename: '',
        content: ''
      });
      // ideally shouldn't need, should be able to utilize component lifecycle
      window.location.reload();
    } catch (err) {
      console.log(err);
    }
  };

  render() {
    let buttonText;
    let handleSubmit;
    if (this.props.update) {
      buttonText = 'Update gist';
      handleSubmit = this.handleUpdateSubmit;
    } else {
      buttonText = 'Create a public gist';
      handleSubmit = this.handleNewSubmit;
    }

    return (
      <div className="gist_form">
        <form action="">
          <FormGroup controlId="formBasicText">
            <FormControl
              className="description"
              type="text"
              value={this.state.description}
              name="description"
              placeholder="Gist description"
              onChange={this.handleChange}
            />
            <FormControl
              className="filename"
              type="text"
              value={this.state.filename}
              name="filename"
              placeholder="Filename including extension"
              onChange={this.handleChange}
            />
            <FormControl
              componentClass="textarea"
              rows={20}
              type="textarea"
              value={this.state.content}
              name="content"
              placeholder="Gist content"
              onChange={this.handleChange}
            />
          </FormGroup>
          <Button bsStyle="primary" onClick={handleSubmit}>
            {buttonText}
          </Button>
        </form>
      </div>
    );
  }
}

export default GistForm;
