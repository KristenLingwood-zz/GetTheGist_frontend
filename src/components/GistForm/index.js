import React, { PureComponent } from 'react';
import { FormControl, FormGroup, Button } from 'react-bootstrap';
import { callAPI } from '../../services/api';
import './gistForm.css';
import AceEditor from 'react-ace';
import 'brace/mode/java';
import 'brace/mode/javascript';
import 'brace/mode/html';
import 'brace/mode/python';
import 'brace/mode/ruby';
import 'brace/mode/markdown';
import 'brace/mode/text';
import 'brace/theme/github';

class GistForm extends PureComponent {
  state = {
    description: '',
    filename: '',
    content: '',
    loading: false,
    fireRedirect: false
  };

  handleChange = e => {
    this.setState({ ...this.state, [e.target.name]: e.target.value });
  };

  handleContentChange = e => {
    console.log('event', e);
    this.setState({ ...this.state, content: e });
  };

  handleNewSubmit = async e => {
    let extension =
      this.state.filename.slice(this.state.filename.indexOf('.') + 1) || '';
    let payload = {
      description: this.state.description,
      filename: this.state.filename,
      content: this.state.content,
      extension: extension
    };
    try {
      await callAPI('post', 'http://localhost:3000/gists', payload);
      window.alert('Thanks for creating a new gist. We got the gist of it!');
      this.setState({
        description: '',
        filename: '',
        content: '',
        fireRedirect: true
      });
      this.props.updateGistList();
    } catch (err) {
      console.log(err);
    }
  };
  handleUpdateSubmit = async e => {
    let payload = {
      description: this.state.description || this.props.foundGist.description,
      filename: this.state.filename || this.props.foundGist.filename,
      content: this.state.content || this.props.foundGist.content,
      extension:
        this.state.filename.slice(this.state.filename.indexOf('.') + 1) ||
        this.props.foundGist.extension
    };
    try {
      let gistID = this.props.gistID;
      let extension =
        this.state.filename.slice(this.state.filename.indexOf('.')) || '';
      let updatedGist = await callAPI(
        'patch',
        `http://localhost:3000/gists/${gistID}`,
        payload
      );
      window.alert('Your gist has been updated!');
      this.setState({
        description: '',
        filename: '',
        content: ''
      });
      this.props.editing();
      this.props.updateGist(updatedGist);
    } catch (err) {
      console.log(err);
    }
  };

  render() {
    let buttonText;
    let handleSubmit;
    if (this.props.update) {
      buttonText = 'Update gist';
      handleSubmit = this.handleUpdateSubmit;
    } else {
      buttonText = 'Create a public gist';
      handleSubmit = this.handleNewSubmit;
    }

    let languageStore = {
      js: 'javascript',
      html: 'html',
      jsp: 'java',
      py: 'python',
      rb: 'ruby',
      md: 'markdown',
      txt: 'text'
    };
    let extension =
      this.state.filename.slice(this.state.filename.indexOf('.') + 1) ||
      'markdown';
    let modeLanguage = languageStore[extension];

    return (
      <div className="gist_form">
        <form>
          <FormGroup controlId="formBasicText">
            <FormControl
              className="description"
              type="text"
              value={this.state.description}
              name="description"
              placeholder="Gist description"
              onChange={this.handleChange}
            />
            <FormControl
              className="filename"
              type="text"
              value={this.state.filename}
              name="filename"
              placeholder="Filename including extension"
              onChange={this.handleChange}
            />
            <AceEditor
              mode={modeLanguage}
              theme="github"
              editorProps={{ $blockScrolling: true }}
              componentClass="textarea"
              rows={20}
              type="textarea"
              value={this.state.content}
              name="content"
              onChange={this.handleContentChange}
            />
          </FormGroup>
          <Button
            bsStyle="primary"
            className="Button --narrow--solid-brand"
            onClick={handleSubmit}
          >
            {buttonText}
          </Button>
        </form>
      </div>
    );
  }
}

export default GistForm;
