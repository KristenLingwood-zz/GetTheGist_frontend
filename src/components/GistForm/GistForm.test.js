import React from 'react';
import GistForm from '.';
import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';

describe('<GistForm />', () => {
  it('matches the snapshot', () => {
    var tree = renderer.create(<GistForm />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<GistForm />, div);
  ReactDOM.unmountComponentAtNode(div);
});
