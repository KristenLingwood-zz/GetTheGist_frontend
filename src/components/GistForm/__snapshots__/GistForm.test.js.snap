// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GistForm /> matches the snapshot 1`] = `
<div
  className="gist_form"
>
  <form>
    <div
      className="form-group"
    >
      <input
        className="description form-control"
        id="formBasicText"
        name="description"
        onChange={[Function]}
        placeholder="Gist description"
        type="text"
        value=""
      />
      <input
        className="filename form-control"
        id="formBasicText"
        name="filename"
        onChange={[Function]}
        placeholder="Filename including extension"
        type="text"
        value=""
      />
      <textarea
        className="form-control"
        id="formBasicText"
        name="content"
        onChange={[Function]}
        placeholder="Gist content"
        rows={20}
        type="textarea"
        value=""
      />
    </div>
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Create a public gist
    </button>
  </form>
</div>
`;
